### PR TITLE
[fix] memory leak in loss reporting

### PR DESCRIPTION
**Pull request by suzyahyah**
Previously `meter_update_dict.update(reduced_loss_dict)` was storing the
loss with the gradient and computation graph. This PR should fix various GPU and CPU oom problems.

### CODE REVIEW

**Comment by suzyahyah** [comment]
not sure why this fails on circle CI, isort and black passed successfully during the commit using editorconfig..

**Comment by vedanuj** [comment]
The circleci test failure is broken on master as well. Not related to this change

**Review by vedanuj — APPROVED** [review]
Thanks for fixing this. Looks good to me!

File: mmf/trainers/core/reporting.py [hunk 1]
```diff
@@ -37,7 +37,10 @@ def update_meter(
 
                 registry.register(loss_key, reduced_loss)
                 meter_update_dict.update({loss_key: reduced_loss})
-                meter_update_dict.update(reduced_loss_dict)
+                for key in reduced_loss_dict:
+                    loss = reduced_loss_dict[key].item()
+                    meter_update_dict[key] = loss
```
Comment (apsdehal) [inline mmf/trainers/core/reporting.py:42]: Do it in single line?

File: mmf/trainers/core/reporting.py [hunk 2]
```diff
@@ -37,7 +37,10 @@ def update_meter(
...
+
             if hasattr(report, "metrics"):
                 meter_update_dict.update(reduced_metrics_dict)
```
Comment (apsdehal) [inline mmf/trainers/core/reporting.py:45]: We should this for metrics as well if metric hasattr item to be foolproof? In rare cases, this may also blow out when metric actually doesn't do .item itself.

**Review by apsdehal — CHANGES_REQUESTED** [review]

File: mmf/trainers/core/reporting.py (hunk 2 above)
Comment (suzyahyah) [reply mmf/trainers/core/reporting.py]: > We should this for metrics as well if metric hasattr item to be foolproof? In rare cases, this may also blow out when metric actually doesn't do .item itself.

Ok, in the latest version I refactored updating `reduced_metrics_dict` and `reduced_loss_dict` to a single update function which checks for `.item()`.

File: mmf/trainers/core/reporting.py [hunk 3]
```diff
@@ -30,14 +30,30 @@ def update_meter(
...
             if hasattr(report, "metrics"):
-                meter_update_dict.update(reduced_metrics_dict)
+                meter_update_dict, _ = self.update_dict(
+                    meter_update_dict, reduced_metrics_dict
+                )
+
             meter.update(meter_update_dict, report.batch_size)
+
+    def update_dict(self, meter_update_dict, values_dict):
+
```
Comment (apsdehal) [inline mmf/trainers/core/reporting.py:48]: is this new line needed?

File: mmf/trainers/core/reporting.py [hunk 4]
```diff
@@ -30,14 +30,30 @@ def update_meter(
...
+        total_val = 0
+        for key, val in values_dict.items():
+            if val.dim() == 1:
```
Comment (apsdehal) [inline mmf/trainers/core/reporting.py:51]: You should first actually check if `dim` attribute is present in val. As for meters, this can be a scalar.

**Review by apsdehal — COMMENTED** [review]
Thank you for making the changes, there are some things that will still fail.

File: mmf/trainers/core/reporting.py (hunk 4 above)
Comment (suzyahyah) [reply mmf/trainers/core/reporting.py]: got it, we now check if `torch.is_tensor` before calling .dim()
